# References

> *ðŸ‘‰ Customize the headers to best reflect the reference content.*

This file lists the scholarly works and software used in this repository. To cite this repository, see the `CITATION.cff` file.

## Papers

- Article citation with DOI.

## Software
 - Bowtie2: Langmead, B., & Salzberg, S. L. (2012). Fast gapped-read alignment with Bowtie 2. Nature methods, 9(4), 357-359. 
 - Samtools: Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., ... & 1000 Genome Project Data Processing Subgroup. (2009). The sequence alignment/map format and SAMtools. bioinformatics, 25(16), 2078-2079.
 - VFDB: Chen, L., Yang, J., Yu, J., Yao, Z., Sun, L., Shen, Y., & Jin, Q. (2005). VFDB: a reference database for bacterial virulence factors. Nucleic acids research, 33(suppl_1), D325-D328.
 - MEGARes: Bonin, N., Doster, E., Worley, H., Pinnell, L. J., Bravo, J. E., Ferm, P., ... & Boucher, C. (2023). MEGARes and AMR++, v3. 0: an updated comprehensive database of antimicrobial resistance determinants and an improved software pipeline for classification using high-throughput sequencing. Nucleic acids research, 51(D1), D744-D752.
 - KMA: Clausen, P. T., Aarestrup, F. M., & Lund, O. (2018). Rapid and precise alignment of raw reads against redundant databases with KMA. BMC bioinformatics, 19, 1-8.
 - CheckM2: Chklovski, A., Parks, D. H., Woodcroft, B. J., & Tyson, G. W. (2023). CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nature Methods, 20(8), 1203-1212.
 - GTDB-tk: Chaumeil, P. A., Mussig, A. J., Hugenholtz, P., & Parks, D. H. (2020). GTDB-Tk: a toolkit to classify genomes with the Genome Taxonomy Database.
 - BLAST: Korf, I., Yandell, M., & Bedell, J. (2003). Blast. " O'Reilly Media, Inc.".
 - BWA: Li, H. (2013). Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv preprint arXiv:1303.3997.
 - GToTree: Lee, M. D. (2019). GToTree: a user-friendly workflow for phylogenomics. Bioinformatics, 35(20), 4162-4164.
 - Atria: Chuan, J., Zhou, A., Hale, L. R., He, M., & Li, X. (2021). Atria: an ultra-fast and accurate trimmer for adapter and quality trimming. GigaByte, 2021, 1-18.
 - Qiime2: Hall, M., & Beiko, R. G. (2018). 16S rRNA gene analysis with QIIME2. In Microbiome analysis: methods and protocols (pp. 113-129). New York, NY: Springer New York.
 - cutadapt: Martin, M. (2011). Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet. journal, 17(1), 10-12.
 - DADA2: Callahan, B. J., McMurdie, P. J., Rosen, M. J., Han, A. W., Johnson, A. J. A., & Holmes, S. P. (2016). DADA2: High-resolution sample inference from Illumina amplicon data. Nature methods, 13(7), 581-583.
 - MAFFT: Katoh, K., Misawa, K., Kuma, K. I., & Miyata, T. (2002). MAFFT: a novel method for rapid multiple sequence alignment based on fast Fourier transform. Nucleic acids research, 30(14), 3059-3066.

