# References

> *ðŸ‘‰ Customize the headers to best reflect the reference content.*

This file lists the scholarly works and software used in this repository. To cite this repository, see the `CITATION.cff` file.

## Papers

- Article citation with DOI.

## Software
 - Bowtie2: Langmead, B., & Salzberg, S. L. (2012). Fast gapped-read alignment with Bowtie 2. Nature methods, 9(4), 357-359. 
 - Samtools: Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., ... & 1000 Genome Project Data Processing Subgroup. (2009). The sequence alignment/map format and SAMtools. bioinformatics, 25(16), 2078-2079.
 - VFDB: Chen, L., Yang, J., Yu, J., Yao, Z., Sun, L., Shen, Y., & Jin, Q. (2005). VFDB: a reference database for bacterial virulence factors. Nucleic acids research, 33(suppl_1), D325-D328.
 - MEGARes: Bonin, N., Doster, E., Worley, H., Pinnell, L. J., Bravo, J. E., Ferm, P., ... & Boucher, C. (2023). MEGARes and AMR++, v3. 0: an updated comprehensive database of antimicrobial resistance determinants and an improved software pipeline for classification using high-throughput sequencing. Nucleic acids research, 51(D1), D744-D752.
 - KMA: Clausen, P. T., Aarestrup, F. M., & Lund, O. (2018). Rapid and precise alignment of raw reads against redundant databases with KMA. BMC bioinformatics, 19, 1-8.
 - CheckM2: Chklovski, A., Parks, D. H., Woodcroft, B. J., & Tyson, G. W. (2023). CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nature Methods, 20(8), 1203-1212.
 - GTDB-tk: Chaumeil, P. A., Mussig, A. J., Hugenholtz, P., & Parks, D. H. (2020). GTDB-Tk: a toolkit to classify genomes with the Genome Taxonomy Database.
 - BLAST: Korf, I., Yandell, M., & Bedell, J. (2003). Blast. " O'Reilly Media, Inc.".
 - BWA: Li, H. (2013). Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv preprint arXiv:1303.3997.
 - GToTree: Lee, M. D. (2019). GToTree: a user-friendly workflow for phylogenomics. Bioinformatics, 35(20), 4162-4164.
 - Atria: Chuan, J., Zhou, A., Hale, L. R., He, M., & Li, X. (2021). Atria: an ultra-fast and accurate trimmer for adapter and quality trimming. GigaByte, 2021, 1-18.
 - Qiime2: Hall, M., & Beiko, R. G. (2018). 16S rRNA gene analysis with QIIME2. In Microbiome analysis: methods and protocols (pp. 113-129). New York, NY: Springer New York.
 - cutadapt: Martin, M. (2011). Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet. journal, 17(1), 10-12.
 - DADA2: Callahan, B. J., McMurdie, P. J., Rosen, M. J., Han, A. W., Johnson, A. J. A., & Holmes, S. P. (2016). DADA2: High-resolution sample inference from Illumina amplicon data. Nature methods, 13(7), 581-583.
 - MAFFT: Katoh, K., Misawa, K., Kuma, K. I., & Miyata, T. (2002). MAFFT: a novel method for rapid multiple sequence alignment based on fast Fourier transform. Nucleic acids research, 30(14), 3059-3066.

## Software Version
- MegaRES v.3.0.
- VFDB v12-06-2023 (complete set)
- KMA v.1.3.23
- Trimmomatic v.0.39.
- FastQC v.0.12.1.
- MultiQC v.1.2.
- Kraken2 v.2.1.3.
- Bracken v.2.9.
- Snakemake v.7.32.4.
- compositions v.2.0-6. (R package)
- vegan v.2.6.4 (R package)
- GUnifrac v.1.8. (R package)
- ggplot2 v.3.4.4. (R package)
- ellipse v.0.5.0. (R package)
- WGCNA v.1.72.5. (R package)
- heatmaply v.1.5.0. (R package)
- lme4 v1.1.35.1. (R package)
- BBDuk (https://sourceforge.net/projects/bbmap/)
- MEGAHIT v.1.2.9.
- Prodigal v.2.6.3.
- Bowtie2 v.7.5.0.
- samtools v.1.19.2. and v.2.6.3.
- CONCOCT v.1.1.0.
- Metabat2 v.2.12.1.
- Das_tools v.1.1.3.
- Checkm2 v.1.0.1.
- GTDB-tk v.2.3.2.
- BLAST v.2.15.0. (nucleotide database v07-31-2023)
- BWA v.0.7.17.
- nlme v.3.1.164. (R package)
- emmeans v.1.9.0. (R package)
- broom v.1.0.5. (R package)
- ggpubr v.0.6.0. (R package)
- gridExtra v.2.3. (R package)
- GToTree v.1.8.4
- HMMER3 v.3.4.
- MUSCLE v.5.1.
- trimAl v.1.4.1.
- FastTree v2.1.11
- iTOL v6
- MAFFT
- DADA2
- Cutadapt
- Qiime2 v.qiime2-2023.7
- Atria v4.0.2
